split http_application_update in two

hakonfam opened the pull request:
Make samples easier to understand, extract common functionality to shared file.

Add separate modem delta update sample.

kyberdin commented on samples/nrf9160/http_update/src/update.c:
```diff
@@ -13,20 +13,32 @@
-static struct k_work	fota_work;
+static const struct device *gpiob;
+static struct gpio_callback gpio_cb;
+static struct k_work fota_work;
```
Was the original alignment not put through clang? I don't see any recent changes in our `.clang-format`. I would think that the formatter would keep this consistent.

hakonfam replied:
Correct, I will do the clang-format now since its being moved anyway.

kyberdin commented on samples/nrf9160/http_update/src/update.c:
```diff
@@ -13,20 +13,32 @@
 
+void (*finished_cb)(void);
+const char *(*host_get)(void);
+const char *(*file_get)(void);
```
Minor: Why not use "get" as the prefix here to make it more statement-like?

hakonfam replied:
In NCS we typically place the verb at the end. So I think the way it is now has the highest probability of being approved in the next round of PRs ;)

kyberdin replied:
Noted, I'll keep that in mind for my own code.

kyberdin commented on samples/nrf9160/http_update/src/update.c:
```diff
@@ -215,18 +222,33 @@ static void modem_configure(void)
 
-	k_work_init(&fota_work, app_dfu_transfer_start);
+	if (finished_cb_in == NULL || host_get_in == NULL || file_get_in == NULL
+	    || two_leds_in ==NULL) {
```
Missing whitespace before `NULL`

kyberdin commented on samples/nrf9160/http_update/modem_delta_update/src/main.c:
```diff
@@ -0,0 +1,116 @@
+		version_read = true;
+	}
+
+	return (strstr(version, FOTA_TEST) != NULL);
```
Minor: I don't think you need parentheses around the logic statement here. Resolve and ignore if you don't care.

kyberdin commented on samples/nrf9160/http_update/modem_delta_update/src/main.c:
```diff
@@ -0,0 +1,116 @@
+	printk("HTTP delta modem update sample started\n");
+
+	printk("Initializing bsdlib\n");
+#if !defined(CONFIG_BSD_LIBRARY_SYS_INIT)
```
References to `bsdlib` were renamed to `libmodem` in the `update.c` file in this PR. Why is it not OK there but fine here?

hakonfam replied:
That is wrong, it should be bsdlib.

Its a bit complicated since we are waiting for a refactor of "bsdlib" which renames it to "libmodem", and the final version of this file (which we use for full modem update) have to use the renamed version :)

kyberdin reviewed (commented):
Initial review.

rlubos commented on samples/nrf9160/http_update/cert/BaltimoreCyberTrustRoot:
```diff
@@ -0,0 +1,81 @@
+/*
```
There are two copies certificate files now (the other one in `http_update/application_update/cert`).

rlubos commented on samples/nrf9160/http_update/application_update/README.rst:
```diff
@@ -40,7 +40,7 @@ This image is automatically generated when building the application.
 Building and running
```
I'm don't think the `samples.rst` will now list these samples given the readme files are nested deeper in the tree:
https://github.com/hakonfam/fw-nrfconnect-nrf-1/blame/http-update/doc/nrf/samples.rst#L37
This could probably be confirmed by some tech writer though.

hakonfam replied:
I will double check, thanks.

hakonfam replied:
fixed

rlubos reviewed (commented):
The overall split of application/modem update into a separate application indeed increased the sample readability at source-code level. But I'm a bit concerned about the current directory structure, i.e. that the actual sample project files are not directly accessible from the sample top directory (`http_update` in this case), this is not something we use to do in NCS and might be confusing for the users.

I'm open for discussion here, but here's the idea I have:
* Add top level `Kconfig` file with an additional `UPDATE_VARIANT` config, specyfing whether to build app update or modem update sample. The configs not used by the currently selected variant could be hidden.
* Add top level `prj.conf` with common configuration and the variant selection (app update by default for instance),
* Add top level `CMakeLists.txt` file sourcing appropriate source files based on configuration and applying variant-specific overlay from the variant directory.
* The sample variant could be switched with an top-level `overlay-*.conf` file

Or if we really need to keep the samples as a separate enitties, group the non-sample directories in the `common` dir (as for instnace OT samples do).

hakonfam commented:
> Or if we really need to keep the samples as a separate enitties, group the non-sample directories in the `common` dir (as for instnace OT samples do).

I suggest we do this. To me one of the biggest motivations for this is the documentation. Will we be able to solve that in a good way if we go with your proposal that we choose in the prj.conf what "sample variant" to run?

hakonfam commented on samples/nrf9160/http_update/src/update.c:
```diff
@@ -0,0 +1,256 @@
+		gpio_pin_configure(dev, DT_GPIO_PIN(DT_ALIAS(led1), gpios),
+				   GPIO_OUTPUT_ACTIVE |
+					   DT_GPIO_FLAGS(DT_ALIAS(led1),
+							 gpios));
```
its clang-format, I'll modify it manually then.

kyberdin replied:
Same for lines [119](https://github.com/nrfconnect/sdk-nrf/blob/7e0bcf318493704d3f107ca14da30d580ee0bde7/samples/nrf9160/http_update/src/update.c#L119) and [150](https://github.com/nrfconnect/sdk-nrf/blob/7e0bcf318493704d3f107ca14da30d580ee0bde7/samples/nrf9160/http_update/src/update.c#L150).

hakonfam replied:
So should we go in and manually modify these, or should we say that whatever clang-format outputs is acceptable?

to me the indendation clang-format outputs (ie the current one) is the most readabel (without extracting to variables, which is a bit hard since these are nested #defines)

hakonfam replied:
To answer mayself, what we should do in situations like this is to modify the code so that clang-format will actually generate good-looking code. Which I have done now :)

rlubos commented:
> > Or if we really need to keep the samples as a separate enitties, group the non-sample directories in the `common` dir (as for instnace OT samples do).
> 
> I suggest we do this. To me one of the biggest motivations for this is the documentation. Will we be able to solve that in a good way if we go with your proposal that we choose in the prj.conf what "sample variant" to run?

I was thinking about top-level `README.rst` with short description of the sample variants with links to the variant-specific readme pages. But I guess `common` subdir can do as well unless someone else has different say.

kyberdin commented on samples/nrf9160/http_update/src/update.c:
```diff
@@ -74,39 +86,25 @@ int cert_provision(void)
-		printk("fota_download_start() failed, err %d\n",
-			retval);
+		printk("fota_download_start() failed, err %d\n", retval);
+		return;
```
Unnecessary `return` when the function ends here anyway.

hakonfam commented:
> I was thinking about top-level `README.rst` with short description of the sample variants with links to the variant-specific readme pages. But I guess `common` subdir can do as well unless someone else has different say.

Me and @lemrey agreed offline that splitting the way it is now (and with `common` subdir) would be a good solution, also it is, as you say, used elsewhere already.

I would suggest we stick with the current split since this makes the samples easiest to understand from a user perspective IMO.

kyberdin commented on samples/nrf9160/http_update/modem_delta_update/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.13.1)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(http_application_update)
```
Project name changed to `http_update`.

hakonfam replied:
will fix

kyberdin commented on samples/nrf9160/http_update/application_update/CMakeLists.txt:
```diff
@@ -9,6 +9,6 @@ cmake_minimum_required(VERSION 3.13.1)
 find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
 project(http_application_update)
```
Project name changed to `http_update`.

hakonfam replied:
will fix

lemrey commented on samples/nrf9160/http_update/application_update/src/main.c:
```diff
@@ -0,0 +1,80 @@
+		return;
+	}
+
+	err = update_sample_init(&(struct update_sample_init_params){
+		.button_pressed = on_button_pressed, .two_leds = two_leds });
```
We can do better than `clang` here

hakonfam replied:
I agree

lemrey commented on samples/nrf9160/http_update/application_update/src/main.c.orig:
```diff
@@ -4,17 +4,24 @@
  * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
  */
 #include <zephyr.h>
+<<<<<<< HEAD
```
This file looks wrong

lemrey commented on samples/nrf9160/http_update/common/include/update.h:
```diff
@@ -0,0 +1,68 @@
+ * @retval false If one LED should be enabled.
+ *
+ */
+typedef bool (*update_sample_two_leds)(void);
```
This could perhaps be renamed to something like `firmware_has_updated()` so that it's easier to understand that it's the criteria with which `update.c` knows whether the update was successful or not.

hakonfam replied:
When we are doing modem delta and full modem update its not so clear with "has_updated" since it will go from two LEDs to one in the second run (which is still technically an "update")

hakonfam replied:
We will also update application_update soon so that you can do an arbitrary number of updates back, making the LED pattern go back and forth between 1 and 2 leds.

lemrey replied:
I think from the point of the implementors (`main.c`), the name should be about what the implementor should return,  rather than what `update.c` will do with the result of the call. This is clearer for the implementors and will keep making sense if, for example, `update.c` decides to display the result in a different way (say, beep). What about a name that suggest that it returns a firmware (binary) "version" or "id" or "variant" e.g. a/b, 1/2.

oyvindronningstad replied:
If you don't want to create a new version/variant concept, I suggest changing it from `bool two_leds` to `int num_leds` or `uint32_t led_mask`.

hakonfam replied:
Changing it to `num_leds`

lemrey commented on samples/nrf9160/http_update/common/src/update.c:
```diff
@@ -0,0 +1,192 @@
+int cert_provision(void)
+{
+	static const char cert[] = {
+#include "../cert/BaltimoreCyberTrustRoot"
```
nit: I think `clang` indented this odd

lemrey commented on samples/nrf9160/http_update/common/src/update.c:
```diff
@@ -0,0 +1,192 @@
+	static const char cert[] = {
+#include "../cert/BaltimoreCyberTrustRoot"
+	};
+	BUILD_ASSERT(sizeof(cert) < KB(4), "Certificate too large");
+	int err;
```
nit: I'd leave a space between that assert and `int err`

lemrey commented on samples/nrf9160/http_update/modem_delta_update/src/main.c:
```diff
@@ -0,0 +1,143 @@
+	}
+
+	err = update_sample_init(&(struct update_sample_init_params){
+		.button_pressed = on_button_pressed, .two_leds = two_leds });
```
ditto

lemrey reviewed (commented):
Good job, I think documentation will be much easier and clearer after this split.
Now that the `main.c` files include the handling of the FOTA callbacks, like we discussed offline, the FOTA logic became clearer, which is good.
There is one file which has a conflict, that needs fixing; the rest are nits :)

hakonfam commented:
We will run CI manually (since we need a branch of the verification repo), so I removed the CI required label and added DNM so that we won't merge until @andvib  has approved (indicating that the CI branch passed)

andvib reviewed (dismissed):
Waiting to run CI

oyvindronningstad commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,29 @@
+	string "Host part of URL to download files."
+
+config SUPPORTED_BASE_VERSION
+	string "Supported base version"
```
Would "Current base version" be correct? Or just "Current version"? Since this is a sample, I feel that a sentence or two of more explanation on some of these configs would be helpful.

hakonfam replied:
We might be in the situation where the user has modem version N installed, but the deltas configured by default is made for version N + 1. In this case we will print the "not supported version, install version N + 1 ... " etc. 

Hence it makes sense to me that this is called "Supported" since we only used it to ensure that the configured deltas supports this version.

Does that make sense? 

I will expand the docs on this.

oyvindronningstad replied:
> Does that make sense?

Haha no, not really. Eagerly awaiting more docs :D.

oyvindronningstad commented on samples/nrf9160/http_update/modem_delta_update/src/main.c:
```diff
@@ -0,0 +1,145 @@
+
+	printk("Current modem firmware version: %s\n", version);
+
+	printk("Press Button 1 for modem delta update\n");
```
```suggestion
	printk("Press Button 1 to perform modem delta update\n");
```

oyvindronningstad commented on samples/nrf9160/http_update/application_update/src/main.c:
```diff
@@ -0,0 +1,82 @@
+	err = fota_download_start(CONFIG_DOWNLOAD_HOST, CONFIG_DOWNLOAD_FILE,
+				  SEC_TAG, apn, 0);
+	if (err != 0) {
+		update_sample_enable_button_irq();
```
Suggestion: Remove this function entirely, always reenable button IRQ, and just ignore `-EALREADY` errors from `fota_download_start`

hakonfam replied:
I'm not sure I understand what you need. I don't want to depend on the specific return values of some API as that is prone to change.

Are you suggesting to leave the button interrupt enabled at all time, and then simply disregard the callback when we have started a download?

oyvindronningstad replied:
We seem to have different views on error codes :D. IMO they are there to be used, and I felt we could make things simpler. Also, `update_sample_enable_button_irq()` doesn't seem to match the rest of the API in terms of level of abstraction. Anyways, it's no big deal for me.

hakonfam replied:
I have broken so many APIs I will never trust one gain ;)

oyvindronningstad commented on samples/nrf9160/http_update/application_update/src/main.c:
```diff
@@ -0,0 +1,82 @@
+
+static bool two_leds(void)
+{
+#if CONFIG_APPLICATION_VERSION == 2
+	return true;
+#else
+	return false;
+#endif
```
```suggestion
return (CONFIG_APPLICATION_VERSION == 2)
```

hakonfam replied:
Changed logic of this function.

oyvindronningstad reviewed (commented):
Mostly cosmetic stuff

hakonfam commented on samples/nrf9160/http_update/common/src/update.c:
```diff
@@ -0,0 +1,193 @@
+	bool exists;
+	uint8_t unused;
+
+	err = modem_key_mgmt_exists(TLS_SEC_TAG,
```
I don't want to introduce changes like this in this PR. Also it seems like bad carma :smile:

oyvindronningstad replied:
Ok, no biggie, was just looking to save 10 LoCs.

hakonfam commented on samples/nrf9160/http_update/common/include/update.h:
```diff
@@ -0,0 +1,68 @@
+ */
+struct update_sample_init_params {
+	/* Callback invoked when the button is pressed */
+	update_sample_button_pressed_cb button_pressed;
```
Went with `upgrade_start` since it follows the "verb at the end" and gives the "update_" namespacing for free.

hakonfam commented on samples/nrf9160/http_update/common/src/update.c:
```diff
@@ -0,0 +1,193 @@
+		return err;
+	}
+
+	err = led_init(params->two_leds());
```
Yes!

kyberdin commented on samples/nrf9160/http_update/modem_delta_update/README.rst:
```diff
@@ -0,0 +1,71 @@
+
+.. include:: /includes/build_and_run.txt
+
+The sample is built as a non-secure firmware image for the nrf9160dk_nrf9160ns build target.
```
```suggestion
The sample is built as a non-secure firmware image for the ``nrf9160dk_nrf9160ns`` build target.
```

kyberdin reviewed (commented):
Mostly minor comments and suggestions.

mswarowsky reviewed (approved):
Tested and verified it works

mswarowsky reviewed (approved):
Retested again

lemrey reviewed (approved):
LGTM, let's invite a tech writer to review the doc changes.

hakonfam commented:
Good point @lemrey, added @FrancescoSer

rlubos commented:
Please add an entry in the changelog for these changes (https://github.com/nrfconnect/sdk-nrf/blob/master/doc/nrf/releases/release-notes-latest.rst).

oyvindronningstad commented on samples/nrf9160/http_update/common/include/update.h:
```diff
@@ -0,0 +1,54 @@
+ * When the button is pressed, its IRQ is disabled automatically.
+ * Use this function to re-enable the button IRQ.
+ */
+void update_sample_enable_button_irq(void);
```
```suggestion
void update_sample_reenable_button_irq(void);
```

hakonfam replied:
Changed to `update_sample_done`

oyvindronningstad commented on samples/nrf9160/http_update/common/src/update.c:
```diff
@@ -0,0 +1,204 @@
+	return 0;
+}
+
+static int led_init(bool two_leds)
```
This hasn't been changed to `num_leds`

oyvindronningstad commented on samples/nrf9160/http_update/common/include/update.h:
```diff
@@ -0,0 +1,54 @@
+	/* Callback invoked when the button is pressed */
+	update_start_cb update_start;
+
+	/* Number of LEDs to enable. Must be 1 or 2. */
```
It would make sense to allow 0 as well IMO.

oyvindronningstad reviewed (changes requested):
Found one error in the `led_init` arg. + cosmetic nits.

FrancescoSer commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,48 @@
+config DOWNLOAD_FILE_BASE_TO_FOTA_TEST
+	string "File to update modem from base to test firmware"
+	help
+	  This sample use the delta patches included in the modem firmware
```
```suggestion
	  This sample uses the delta patches included in the modem firmware
```

FrancescoSer commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,48 @@
+	string "File to update modem from base to test firmware"
+	help
+	  This sample use the delta patches included in the modem firmware
+	  releases. These delta patches only works for the modem firmware
```
```suggestion
	  releases. These delta patches only work for the modem firmware
```

FrancescoSer commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,48 @@
+config DOWNLOAD_FILE_FOTA_TEST_TO_BASE
+	string "File to update modem from test firmware base"
+	help
+	  This sample use the delta patches included in the modem firmware
```
```suggestion
	  This sample uses the delta patches included in the modem firmware
```

FrancescoSer commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,48 @@
+	string "File to update modem from test firmware base"
+	help
+	  This sample use the delta patches included in the modem firmware
+	  releases. These delta patches only works for the modem firmware
```
```suggestion
	  releases. These delta patches only work for the modem firmware
```

FrancescoSer commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,48 @@
+config SUPPORTED_BASE_VERSION
+	string "Supported base version"
+	help
+	  This sample use the delta patches included in the modem firmware
```
```suggestion
	  This sample uses the delta patches included in the modem firmware
```

FrancescoSer commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,48 @@
+	string "Supported base version"
+	help
+	  This sample use the delta patches included in the modem firmware
+	  releases. These delta patches only works for the modem firmware
```
```suggestion
	  releases. These delta patches only work for the modem firmware
```

FrancescoSer commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,48 @@
+	help
+	  This sample use the delta patches included in the modem firmware
+	  releases. These delta patches only works for the modem firmware
+	  version which they are bundled with. This option specifies file part
```
```suggestion
	  version which they are bundled with. This option specifies the file part
```

FrancescoSer commented on samples/nrf9160/http_update/modem_delta_update/Kconfig:
```diff
@@ -0,0 +1,48 @@
+	help
+	  This sample use the delta patches included in the modem firmware
+	  releases. These delta patches only works for the modem firmware
+	  version which they are bundled with. This option specifies file part
```
```suggestion
	  version which they are bundled with. This option specifies the file part
```

FrancescoSer reviewed (approved):
Reviewed. Also added a pair of notes on one of the Kconfig files.
